gh-72971: Clarify the special no-TypeError behavior for equality (#110729)
python/cpython: Doc/reference/datamodel.rst

   the left operand's method has priority.  Virtual subclassing is
   not considered.

>    When no method returns any value other than NotImplemented, comparisons raise
>    a :exc:`TypeError`. However, the ``==`` and ``!=`` operators will not raise an
>    exception in that case, and will instead fall back to ``is`` and ``is not``,
>    respectively. That way, even though ``object().__eq__(object())`` returns
>    NotImplemented, ``object() == object()`` returns False.
> 

Reviewer: I would prefer a shorter version:
```suggestion
   When no appropriate method returns any value other than ``NotImplemented``, the
   ``==`` and ``!=`` operators will fall back to ``is`` and ``is not``, respectively.
```

Author: Ok, but I will add that other comparison operators raise a TypeError, so that no blur remains.

Reviewer: I think [it's redundant](https://docs.python.org/3/reference/datamodel.html?#special-method-names): "Except where mentioned, attempts to execute an operation raise an exception when no appropriate method is defined (typically [AttributeError](https://docs.python.org/3/library/exceptions.html#AttributeError) or [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError))."  That's probably why similar cases were not mentioned e.g. in [Emulating numeric types](https://docs.python.org/3/reference/datamodel.html?#emulating-numeric-types).

So, count me -1 on this.

Author: Fair enough.

   rather, :meth:`__lt__` and :meth:`__gt__` are each other's reflection,
   :meth:`__le__` and :meth:`__ge__` are each other's reflection, and
   :meth:`__eq__` and :meth:`__ne__` are their own reflection.
   If the operands are of different types, and the right operand's type is
   a direct or indirect subclass of the left operand's type,
   the reflected method of the right operand has priority, otherwise
   the left operand's method has priority.  Virtual subclassing is
   not considered.

>    When no appropriate method returns any value other than :data:`!NotImplemented`, the

Reviewer 2: Restore link, since this is the first ref in this paragraph.

```suggestion
   When no appropriate method returns any value other than :data:`NotImplemented`, the
```
